В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекомендую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [2]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [3]:
!pip install statsmodels --quiet

In [4]:
import numpy as np
from statsmodels.stats.weightstats import ztest

In [5]:
heights = np.array(heights)

In [6]:
mean_height = 170
alpha = 0.05

In [7]:
mean_height = np.mean(heights)

In [8]:
print(f"Середній зріст: {mean_height}")

Середній зріст: 170.6969696969697


In [24]:
np.std(heights), np.mean(heights)

(np.float64(4.648133209409457), np.float64(170.6969696969697))

In [26]:
z_stat, p_value = ztest(heights, value=mean_height)

In [27]:
ztest_Score, p_value = ztest(heights, value = mean_height, alternative='larger')

In [31]:
z_stat, ztest_Score, p_value

(np.float64(0.0), np.float64(0.0), np.float64(0.5))

In [32]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


В завданні середній зріст студентів 170 см. При розрахунку на даних вибірки середній зріст студентів 170,69 см. Що майже тотожне заданому середньому значенню.
Стандартне відхилення 4,648 см. Що є реальним, якщо подивитись на вибірку.
Альфа 0,05.
Приймаємо нульову гіпотезу

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [14]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [33]:
group_a_scores = np.array(group_a_scores)
group_b_scores = np.array(group_b_scores)

In [34]:
np.array(group_a_scores)

array([78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06])

In [35]:
np.array(group_b_scores)

array([81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19])

In [36]:
np.mean(group_a_scores), np.mean(group_b_scores)

(np.float64(77.08125), np.float64(74.73875000000001))

In [37]:
alpha = 0.05

In [40]:
ztest(group_a_scores, group_b_scores, alternative='two-sided')

(np.float64(2.357354864751502), np.float64(0.01840564855252801))

In [41]:
ztest(group_a_scores, group_b_scores, value=0)

(np.float64(2.357354864751502), np.float64(0.01840564855252801))

Статистика для двох вибірок z-критерію становить 2.3573, а відповідне p-value дорівнює 0.0184.

Оскільки p-value < 0.05, ми відкидаємо нульову гіпотезу. Тобто, середні оцінки значно відрізняються в групи, яка отримала новий навчальний курс, від групи, яка продовжила за стандартним курсом.

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [44]:
from scipy.stats import ttest_ind

In [49]:
group_a_subset = group_a_scores[:5]

In [51]:
stat, p_value = ttest_ind(group_a_subset, group_b_scores, equal_var=True)

print(f"t_stat: {stat:.3f}")
print(f"p_value: {p_value:.3f}")

t_stat: 0.817
p_value: 0.419


In [52]:
ttest_ind(group_a_subset, group_b_scores, axis=0, equal_var=True, nan_policy='propagate', permutations=None, random_state=None, alternative='two-sided', trim=0, method=None, keepdims=False)

<ipython-input-52-b706e8a22672>:1: DeprecationWarning: Arguments {'random_state', 'permutations'} are deprecated, whether passed by position or keyword. They will be removed in SciPy 1.17.0. Use ``method`` to perform a permutation test.
  ttest_ind(group_a_subset, group_b_scores, axis=0, equal_var=True, nan_policy='propagate', permutations=None, random_state=None, alternative='two-sided', trim=0, method=None, keepdims=False)


TtestResult(statistic=np.float64(0.8167951030410814), pvalue=np.float64(0.4185479477535733), df=np.float64(43.0))

На основі проведеного т тесту можна сказати, що середні значення двох груп  відрізняються на рівні 0.817, що є меншою різницею від попереднього тесту.
p_value становить 0.419  і є більший за альфа. Нульова гіпотеза не може бути відхилена.
Тобто, на основі вибірки 5 оцінок з нового навчального курсу не можна зробити висновок, що цей новий курс є ефективнішим. Замала вибірка для порівняння з іншою більшою вибіркою оцінок.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [53]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [54]:
np.mean(before_training), np.mean(after_training)

(np.float64(50.885999999999996), np.float64(54.03600000000001))

In [55]:
z_stat, p_value = ztest(before_training, after_training)

In [56]:
ztest(before_training, after_training, value=0)

(np.float64(-1.3668871347639466), np.float64(0.17166068512470822))

Статистика для двох вибірок z-критерію становить -1.36688, а відповідне p-value дорівнює 0.17166.
Оскільки p-value > 0.05, у нас є достатньо підтвердженб щоб підтвердити нульову гіпотезу. Іншими словами, тренінг не вплинув на продуктивність працівників.